web/modules/custom/pineapple_general/pineapple_general.module

```diff
@@ -108,6 +93,56 @@ function pineapple_general_form_alter(&$form, \Drupal\Core\Form\FormStateInterfa
   }
 }
 
+/**
+ * Implements hook_field_widget_WIDGET_TYPE_form_alter().
+ */
+function pineapple_general_field_widget_entity_reference_paragraphs_form_alter(&$element, \Drupal\Core\Form\FormStateInterface $form_state, $context) {
+  $user_roles = \Drupal::currentUser()->getRoles();
+   /* For each item in the widget array, if paragraph type value is equal to
+   image_paragraph. */
+ if (isset($element['subform']['field_p_c_component']) && !in_array('administrator', $user_roles)) {
+   foreach ($element['subform']['field_p_c_component']['widget'] as &$paragraph) {
+     if (isset($paragraph['#paragraph_type'])) {
+       if ($paragraph['#paragraph_type'] == 'image_paragraph' || $paragraph['#paragraph_type'] == 'image_grid') {
+         if (isset($paragraph['subform']['field_p_ip_hero_image']) || isset($paragraph['subform']['field_p_ig_item'])) {
+           /* For each $parent in the field_parents array, if the parent is in an
+           article or recipe content type, restrict access to the hero image field. */
+           $paragraph['subform']['field_p_ip_hero_image']['#access'] = FALSE;
+
+           foreach($paragraph['subform']['field_p_ig_item']['widget'] as &$paragraph_field) {
+             if (isset($paragraph_field['subform']['field_p_gi_link'])) {
+               /* For each $parent in the field_parents array, if the parent is in an
+               article or recipe content type, restrict access to the hero image field. */
+               $paragraph_field['subform']['field_p_gi_link']['#access'] = FALSE;
+               foreach ($paragraph_field['subform']['field_p_gi_image']['widget'] as &$image_paragraph_items) {
+                 if (isset($image_paragraph_items['subform']['field_p_ip_hero_image'])) {
+                   $image_paragraph_items['subform']['field_p_ip_hero_image']['#access'] = FALSE;
+                 }
+               }
+             }
+           }
+         }
+       }
+     }
+   }
+ }
+
+  if ($element['#paragraph_type'] == 'components') {
+    $removed_links = [
+      'add_more_button_call_to_action',
+      'add_more_button_slideshow',
+      'add_more_button_content_spotlight'
+    ];
+    if (isset($element['subform']['field_p_c_component']['widget']['add_more']['operations']['#links']) && !in_array('administrator', $user_roles)) {
+      foreach ($element['subform']['field_p_c_component']['widget']['add_more']['operations']['#links'] as $key => $add_more) {
+        if (in_array($key, $removed_links)) {
+          unset($element['subform']['field_p_c_component']['widget']['add_more']['operations']['#links'][$key]);
+        }
```

Comment:
You should turn ```$element['subform']['field_p_c_component']['widget']['add_more']['operations']['#links']``` into a variable so that it's a little shorter and would make the code easier to read.

Reply:
Done